packages/cli/src/ui/contexts/KeypressContext.tsx

```diff
@@ -19,427 +14,733 @@ import {
+          keypressHandler({
+            name: '',
+            ctrl: false,
+            meta: false,
+            shift: false,
+            paste: true,
+            sequence: buffer,
+          });
+          buffer = null;
+        } else if (key != null) {
+          buffer = key.sequence;
+          timeoutId = setTimeout(() => bufferer.next(null), PASTE_TIMEOUT);
+        }
+      } else if (buffer !== null) {
+        buffer += key.sequence;
+        timeoutId = setTimeout(() => bufferer.next(null), PASTE_TIMEOUT);
+      } else {
+        keypressHandler(key);
       }
     }
+  })();
+  bufferer.next(); // prime the generator so it starts listening.
 
-    const kittyKeyCodeToName: { [key: number]: string } = {
-      [CHAR_CODE_ESC]: 'escape',
-      [KITTY_KEYCODE_TAB]: 'tab',
-      [KITTY_KEYCODE_BACKSPACE]: 'backspace',
-      [KITTY_KEYCODE_ENTER]: 'return',
-      [KITTY_KEYCODE_NUMPAD_ENTER]: 'return',
-    };
-
-    const name = kittyKeyCodeToName[keyCode];
-    if (name) {
-      return {
-        key: {
-          name,
-          ctrl,
-          meta: alt,
-          shift,
-          paste: false,
-          sequence: buffer.slice(0, m[0].length),
-          kittyProtocol: true,
-        },
-        length: m[0].length,
-      };
-    }
+  return (key: Key | null) => {
+    clearTimeout(timeoutId);
+    bufferer.next(key);
+  };
+}
 
-    // Ctrl+letters and Alt+letters
-    if (
-      (ctrl || alt) &&
-      keyCode >= 'a'.charCodeAt(0) &&
-      keyCode <= 'z'.charCodeAt(0)
-    ) {
-      const letter = String.fromCharCode(keyCode);
-      return {
-        key: {
-          name: letter,
-          ctrl,
-          meta: alt,
-          shift,
-          paste: false,
-          sequence: buffer.slice(0, m[0].length),
-          kittyProtocol: true,
-        },
-        length: m[0].length,
-      };
+/**
+ * Buffers paste events between paste-start and paste-end sequences.
+ * Will flush the buffer if no data is received for PASTE_TIMEOUT ms or
+ * when a null key is received.
+ */
+function bufferPaste(
+  keypressHandler: KeypressHandler,
+): (key: Key | null) => void {
+  let timeoutId: NodeJS.Timeout;
+  const bufferer = (function* (): Generator<void, void, Key | null> {
+    // This is null when not in paste mode
+    let buffer: string | null = null;
+    while (true) {
+      const key = yield;
+      if (key == null || key.name === 'paste-end') {
+        if (buffer !== null) {
+          keypressHandler({
+            name: '',
+            ctrl: false,
+            meta: false,
+            shift: false,
+            paste: true,
+            sequence: buffer,
+          });
+        }
+        buffer = null;
+      } else if (key.name === 'paste-start') {
+        buffer = '';
+        timeoutId = setTimeout(() => bufferer.next(null), PASTE_TIMEOUT);
+      } else if (buffer !== null) {
+        buffer += key.sequence;
+        timeoutId = setTimeout(() => bufferer.next(null), PASTE_TIMEOUT);
```

Comment:
sorry dumb question. should we be cancelling the previous timeout before setting a new one?